### PR TITLE
#130341273 Added JWT authentication for login

### PR DESCRIPTION
**Pull request by michael-alade**
### What does this PR do?

This PR adds a feature where users are given a JWT(JSON web token) after a successful login.
#### Description of Task to be completed?

It completes the task for user login when the (/api/auth/login) endpoint is sent a POST method with the user's details.
#### How should this be manually tested?

To test the feature, install postman and try this API endpoint with a POST method `http://localhost:PORT/api/auth/login`
##### Add the following form fields
- email
- password

Success response

``` javascript
{
"success": true,
"message": "Successfully logged in"
"token":"eyJhbGciOiJIUzI1NiIsInR5porpwirwcCI6IkpXVCJ9.eyJfaWQiOiI1N2YyZjBhYTgxOTA5NTNmNTgwMDAwMDEiLCJpYXQiOjE0NzU1difroeirorkwoieMzkxMTQsImV4cCI6MTQ3NTU0OTE5NH0.rCJdm6tvHa9v1XX9o3JXlTnEna1mVToEmh5vPH0V-O0"
}
```
#### Background context

JSON Web Token (JWT) is an open standard (RFC 7519) that defines a compact and self-contained way for securely transmitting information between parties as a JSON object. This information can be verified and trusted because it is digitally signed. JWTs can be signed using a secret (with the HMAC algorithm) or a public/private key pair using RSA.

JSON Web Tokens consist of three parts separated by dots (.), which are:
- Header
- Payload
- Signature
#### Pivotal tracker stories?
#130341273

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -4,18 +4,21 @@
...
   res.redirect('/chooseavatars');
 };
 
 /**
  * Show login form
  */
-exports.signin = function(req, res) {
+exports.signin = function (req, res) {
   if (!req.user) {
     res.redirect('/#!/signin?error=invalid');
```
Comment (danielonodje) [inline app/controllers/users.js:23]: is the error passed in a flash message?

File: app/controllers/users.js [hunk 2]
```diff
@@ -37,116 +40,175 @@ exports.signup = function(req, res) {
...
   res.redirect('/');
 };
 
 /**
  * Session
  */
+
+
+exports.getAllUser = function (req, res, next) {
+  User.find({}, function (err, user) {
```
Comment (danielonodje) [inline app/controllers/users.js:54]: should this be findAll?

File: app/controllers/users.js [hunk 3]
```diff
@@ -37,116 +40,175 @@ exports.signup = function(req, res) {
...
-    .exec(function(err, user) {
-      if (user.avatar !== undefined) {
-        res.redirect('/#!/');
-      } else {
-        res.redirect('/#!/choose-avatar');
-      }
-    });
+      .exec(function (err, user) {
+        if (user.avatar !== undefined) {
+          res.redirect('/#!/');
```
Comment (danielonodje) [inline app/controllers/users.js:86]: what does this do?

File: app/controllers/users.js [hunk 4]
```diff
@@ -37,116 +40,175 @@ exports.signup = function(req, res) {
...
+      token = user.generateJwt();
+      res.status(200);
+      res.json({
+        success: true,
+        message: 'Successfully logged on.',
+        token: token
+      });
+    } else {
+      // If user is not found
+      res.status(401).json(info);
```
Comment (danielonodje) [inline app/controllers/users.js:116]: nice. didn't know you could do this

File: app/controllers/validators.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+
+exports.validateEmail = function (email) {
+    var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (danielonodje) [inline app/controllers/validators.js:3]: I think there's definitely a shorter way to validate this

**Review by danielonodje — COMMENTED** [review]
please don't use process.env in code. all config vars should be available via the config object as seen in server.js

**Comment by danielonodje** [comment]
We (the team) need to talk about password standards as it's a security issue that affects the whole project.

File: test/routes/login.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
+var request = require('supertest'),
```
Comment (andela-oadebayo) [inline test/routes/login.js:1]: @andela-akolawole 
Please visit this site https://httpstatuses.com/, and fix all status code issues in your test.

File: test/routes/login.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        };
+
+        request(url)
+            .post('/api/auth/login')
+            .send(logInUser)
+            .expect('Content-Type', 'application/json; charset=utf-8')
+            .expect(401)
+            .end(function (err, res) {
+                if (err) throw err;
+                res.body.should.have.property('success');
```
Comment (andela-oadebayo) [inline test/routes/login.js]: @andela-akolawole 
I feel the success property here is very confusing, the test states `should return error if user not found in database`, so i believe you need an error property there.

**Comment by andela-kakpobome** [comment]
@andela-akolawole is there a reason why your branch naming does not follow the standard set in the wiki?

**Comment by michael-alade** [comment]
@andela-kakpobome this branch naming convention was used before the update to the wiki

**Pull request merged**